#18108번

ICPC Bangkok Regional에 참가하기 위해 수완나품 국제공항에 막 도착한 팀 레드시프트 일행은 눈을 믿을 수 없었다. 공항의 대형 스크린에 올해가 2562년이라고 적혀 있던 것이었다.

불교 국가인 태국은 불멸기원(佛滅紀元), 즉 석가모니가 열반한 해를 기준으로 연도를 세는 불기를 사용한다. 반면, 우리나라는 서기 연도를 사용하고 있다. 불기 연도가 주어질 때 이를 서기 연도로 바꿔 주는 프로그램을 작성하시오.

In [ ]:
century = int(input())
print(century - 543)

2541
1998


#10430번

(A+B)%C는 ((A%C) + (B%C))%C 와 같을까?

(A×B)%C는 ((A%C) × (B%C))%C 와 같을까?

세 수 A, B, C가 주어졌을 때, 위의 네 가지 값을 구하는 프로그램을 작성하시오.

In [ ]:
[a, b, c] = input().split()

a = int(a)
b = int(b)
c = int(c)


print((a+b)%c)
print(((a%c)+(b%c))%c)
print((a*b)%c)
print(((a%c)*(b%c))%c)

5 8 4
1.0
1.0
0.0
0.0


# 2588번

(세 자리 수) × (세 자리 수)는 다음과 같은 과정을 통하여 이루어진다.

472 ... 1 </br>
385 x ... 2</br>
--------------</br>
2360 ... 3</br>
3776 ... 4</br>
1416 ... 5</br>
--------------</br>
181720

(1)과 (2)위치에 들어갈 세 자리 자연수가 주어질 때 (3), (4), (5), (6)위치에 들어갈 값을 구하는 프로그램을 작성하시오.

In [ ]:
a = int(input())
b = input()
b_1 = int(b[2])
b_2 = int(b[1])
b_3 = int(b[0])
b = int(b)

print(a*b_1)
print(a*b_2)
print(a*b_3)
print(a*b)

472
385
2360
3776
1416
181720


# 1330번

두 정수 A와 B가 주어졌을 때, A와 B를 비교하는 프로그램을 작성하시오.

In [ ]:
[a, b] = input().split()
a = int(a)
b = int(b)

if (a > b):
  print('>')
elif (a < b):
  print('<')
else:
  print('==')

1 2
<


# 9498번

시험 점수를 입력받아 90 ~ 100점은 A, 80 ~ 89점은 B, 70 ~ 79점은 C, 60 ~ 69점은 D, 나머지 점수는 F를 출력하는 프로그램을 작성하시오.

In [ ]:
score = int(input())

if (score >= 90):
  print('A')
elif (score >= 80):
  print('B')
elif (score >= 70):
  print('C')
elif (score >= 60):
  print('D')
else:
  print('F')

70
C


# 2753번

연도가 주어졌을 때, 윤년이면 1, 아니면 0을 출력하는 프로그램을 작성하시오.

윤년은 연도가 4의 배수이면서, 100의 배수가 아닐 때 또는 400의 배수일 때이다.

예를 들어, 2012년은 4의 배수이면서 100의 배수가 아니라서 윤년이다. 1900년은 100의 배수이고 400의 배수는 아니기 때문에 윤년이 아니다. 하지만, 2000년은 400의 배수이기 때문에 윤년이다.

In [ ]:
year = int(input())

if (year % 400 == 0):
  print(1)
elif (year % 100 == 0):
  print(0)
elif (year % 4 == 0):
  print(1)
else:
  print(0)

2012
1


# 14681번

 수학 문제 중 하나는 주어진 점이 어느 사분면에 속하는지 알아내는 것이다. 사분면은 아래 그림처럼 1부터 4까지 번호를 갖는다. "Quadrant n"은 "제n사분면"이라는 뜻이다.

예를 들어, 좌표가 (12, 5)인 점 A는 x좌표와 y좌표가 모두 양수이므로 제1사분면에 속한다. 점 B는 x좌표가 음수이고 y좌표가 양수이므로 제2사분면에 속한다.

점의 좌표를 입력받아 그 점이 어느 사분면에 속하는지 알아내는 프로그램을 작성하시오. 단, x좌표와 y좌표는 모두 양수나 음수라고 가정한다.

In [ ]:
x = int(input())
y = int(input())

if (x > 0):
  if (y > 0):
    print(1)
  elif(y < 0):
    print(4)
elif(x < 0):
  if (y > 0):
    print(2)
  elif(y < 0):
    print(3)

12
5
1


# 2884번

상근이는 매일 아침 알람을 듣고 일어난다. 알람을 듣고 바로 일어나면 다행이겠지만, 항상 조금만 더 자려는 마음 때문에 매일 학교를 지각하고 있다.

상근이는 모든 방법을 동원해보았지만, 조금만 더 자려는 마음은 그 어떤 것도 없앨 수가 없었다.

이런 상근이를 불쌍하게 보던, 창영이는 자신이 사용하는 방법을 추천해 주었다.

바로 "45분 일찍 알람 설정하기"이다.

이 방법은 단순하다. 원래 설정되어 있는 알람을 45분 앞서는 시간으로 바꾸는 것이다. 어차피 알람 소리를 들으면, 알람을 끄고 조금 더 잘 것이기 때문이다. 이 방법을 사용하면, 매일 아침 더 잤다는 기분을 느낄 수 있고, 학교도 지각하지 않게 된다.

현재 상근이가 설정한 알람 시각이 주어졌을 때, 창영이의 방법을 사용한다면, 이를 언제로 고쳐야 하는지 구하는 프로그램을 작성하시오.

In [ ]:
[H, M] = input().split()
H = int(H)
M = int(M)

M = M - 45
if (M < 0):
  H = H - 1
  M = M + 60
if (H < 0):
  H = H + 24

print(H, M)

10 10
9 25


# 2525번

KOI 전자에서는 건강에 좋고 맛있는 훈제오리구이 요리를 간편하게 만드는 인공지능 오븐을 개발하려고 한다. 인공지능 오븐을 사용하는 방법은 적당한 양의 오리 훈제 재료를 인공지능 오븐에 넣으면 된다. 그러면 인공지능 오븐은 오븐구이가 끝나는 시간을 분 단위로 자동적으로 계산한다. 

또한, KOI 전자의 인공지능 오븐 앞면에는 사용자에게 훈제오리구이 요리가 끝나는 시각을 알려 주는 디지털 시계가 있다. 

훈제오리구이를 시작하는 시각과 오븐구이를 하는 데 필요한 시간이 분단위로 주어졌을 때, 오븐구이가 끝나는 시각을 계산하는 프로그램을 작성하시오.

In [ ]:
[current_H, current_M] = [int(i) for i in input().split()]
require_M = int(input())

require_H = require_M // 60
require_M %= 60

complete_M = current_M + require_M
complete_H = current_H + require_H

if (complete_M >= 60):
  complete_M -= 60
  complete_H += 1
if (complete_H >= 24):
  complete_H -= 24

print(complete_H, complete_M)

23 48
25
0 13


1에서부터 6까지의 눈을 가진 3개의 주사위를 던져서 다음과 같은 규칙에 따라 상금을 받는 게임이 있다. 

1. 같은 눈이 3개가 나오면 10,000원+(같은 눈)×1,000원의 상금을 받게 된다. 
2. 같은 눈이 2개만 나오는 경우에는 1,000원+(같은 눈)×100원의 상금을 받게 된다. 
3. 모두 다른 눈이 나오는 경우에는 (그 중 가장 큰 눈)×100원의 상금을 받게 된다.  

예를 들어, 3개의 눈 3, 3, 6이 주어지면 상금은 1,000+3×100으로 계산되어 1,300원을 받게 된다. 또 3개의 눈이 2, 2, 2로 주어지면 10,000+2×1,000 으로 계산되어 12,000원을 받게 된다. 3개의 눈이 6, 2, 5로 주어지면 그중 가장 큰 값이 6이므로 6×100으로 계산되어 600원을 상금으로 받게 된다.

3개 주사위의 나온 눈이 주어질 때, 상금을 계산하는 프로그램을 작성 하시오.

In [ ]:
[a, b, c] = [int(i) for i in input().split()]

prize = 0

if (a == b):
  if (b == c):
    prize = 10000 + a * 1000
  else:
    prize = 1000 + a * 100
elif (a == c):
  prize = 1000 + a * 100
elif (b == c):
  prize = 1000 + b * 100
else:
  prize = max(a, b, c) * 100

print(prize)

6 2 5
600


# 2739번

N을 입력받은 뒤, 구구단 N단을 출력하는 프로그램을 작성하시오. 출력 형식에 맞춰서 출력하면 된다.

In [ ]:
N = int(input())

for i in range(1, 10):
  print(f'{N} * {i} = {N*i}')

3
3 * 1 = 3
3 * 2 = 6
3 * 3 = 9
3 * 4 = 12
3 * 5 = 15
3 * 6 = 18
3 * 7 = 21
3 * 8 = 24
3 * 9 = 27


# 10950번

두 정수 A와 B를 입력받은 다음, A+B를 출력하는 프로그램을 작성하시오.

In [ ]:
T = int(input())

for i in range(T):
  print(sum(map(int, input().split())))

1
10 5
15


# 8393번

n이 주어졌을 때, 1부터 n까지 합을 구하는 프로그램을 작성하시오.

In [ ]:
n = int(input())

sum = 0
for i in range(n+1):
  sum += i
print(sum)

10
55


# 15552번 

Python을 사용하고 있다면, input 대신 sys.stdin.readline을 사용할 수 있다. 단, 이때는 맨 끝의 개행문자까지 같이 입력받기 때문에 문자열을 저장하고 싶을 경우 .rstrip()을 추가로 해 주는 것이 좋다.

In [ ]:
import sys

T = int(input())
for i in range(T):
  print(sum(map(int, sys.stdin.readline().split())))

# 2741번

자연수 N이 주어졌을 때, 1부터 N까지 한 줄에 하나씩 출력하는 프로그램을 작성하시오.

In [ ]:
n = int(input())
for i in range(1, n+1):
  print(i)

5
1
2
3
4
5


# 2742번

자연수 N이 주어졌을 때, N부터 1까지 한 줄에 하나씩 출력하는 프로그램을 작성하시오.

In [ ]:
n = int(input())

for i in range(1, n+1)[::-1]:
  print(i)

5
5
4
3
2
1


# 11021번

두 정수 A와 B를 입력받은 다음, A+B를 출력하는 프로그램을 작성하시오.

In [ ]:
t = int(input())

for i in range(1, t+1):
  result = sum(map(int, input().split()))
  print(f'Case #{i}: {result}')

# 11022번

두 정수 A와 B를 입력받은 다음, A+B를 출력하는 프로그램을 작성하시오.

In [ ]:
t = int(input())

for i in range(1, t+1):
  [a, b] = map(int, input().split())
  print(f'Case #{i}: {a} + {b} = {a+b}')

# 2438번

첫째 줄에는 별 1개, 둘째 줄에는 별 2개, N번째 줄에는 별 N개를 찍는 문제

In [ ]:
n = int(input())

for i in range(n):
  print('*'*(i+1))

5
*
**
***
****
*****


# 2439번

첫째 줄에는 별 1개, 둘째 줄에는 별 2개, N번째 줄에는 별 N개를 찍는 문제

하지만, 오른쪽을 기준으로 정렬한 별(예제 참고)을 출력하시오.

In [ ]:
n = int(input())

for i in range(1, n+1):
  stars = ' ' * (n - i) + '*' * i
  print(stars)

5
    *
   **
  ***
 ****
*****


# 10871번

정수 N개로 이루어진 수열 A와 정수 X가 주어진다. 이때, A에서 X보다 작은 수를 모두 출력하는 프로그램을 작성하시오.

In [ ]:
[n, x] = [int(i) for i in input().split()]

arr = list(map(int, input().split()))

smaller_elements = ''

for i in range(n):
  if (arr[i] < x):
    smaller_elements += (' ' + str(arr[i]))

print(smaller_elements[1:])

10 5
1 10 4 9 2 3 8 5 7 6
1 4 2 3


# 10952번

두 정수 A와 B를 입력받은 다음, A+B를 출력하는 프로그램을 작성하시오.

In [ ]:
while(True):
  [a, b] = [int(i) for i in input().split()]
  
  if (a == 0 and b == 0):
    break

  print(a + b)  

1 1
2
2 3
5
3 4
7
9 8
17
5 2
7
0 0


# 10951번

두 정수 A와 B를 입력받은 다음, A+B를 출력하는 프로그램을 작성하시오.

In [ ]:
import sys

line = sys.stdin.readline()
while(line != ''):
  [a, b] = [int(i) for i in line.split()]
  print(a + b)
  line = sys.stdin.readline()

# 1110번

0보다 크거나 같고, 99보다 작거나 같은 정수가 주어질 때 다음과 같은 연산을 할 수 있다. 먼저 주어진 수가 10보다 작다면 앞에 0을 붙여 두 자리 수로 만들고, 각 자리의 숫자를 더한다. 그 다음, 주어진 수의 가장 오른쪽 자리 수와 앞에서 구한 합의 가장 오른쪽 자리 수를 이어 붙이면 새로운 수를 만들 수 있다. 다음 예를 보자.

26부터 시작한다. 2+6 = 8이다. 새로운 수는 68이다. 6+8 = 14이다. 새로운 수는 84이다. 8+4 = 12이다. 새로운 수는 42이다. 4+2 = 6이다. 새로운 수는 26이다.

위의 예는 4번만에 원래 수로 돌아올 수 있다. 따라서 26의 사이클의 길이는 4이다.

N이 주어졌을 때, N의 사이클의 길이를 구하는 프로그램을 작성하시오.

In [ ]:
c = 1

n = int(input())
origin_n = n


while (True):

  n_10 = n // 10
  n_1 = n % 10

  new_num = n_1 * 10 + (n_10 + n_1) % 10
  if(origin_n == new_num):
    break
  else:
    c += 1
    n = new_num

print(c)

0
1


# 10818번

N개의 정수가 주어진다. 이때, 최솟값과 최댓값을 구하는 프로그램을 작성하시오.

In [ ]:
n = int(input())
arr = list(map(int, input().split()))

min = 1000001
max = -1000001

for i in arr:
  if (i < min):
    min = i
  if (i > max):
    max = i

print(min, max)

5
20 10 35 30 7
7 35


# 2562번

9개의 서로 다른 자연수가 주어질 때, 이들 중 최댓값을 찾고 그 최댓값이 몇 번째 수인지를 구하는 프로그램을 작성하시오.

예를 들어, 서로 다른 9개의 자연수

3, 29, 38, 12, 57, 74, 40, 85, 61

이 주어지면, 이들 중 최댓값은 85이고, 이 값은 8번째 수이다.

In [ ]:
arr = []
for _ in range(9):
  arr.append(int(input()))


max = arr[0]
index = 1

for i, n in enumerate(arr):
  if (n > max):
    max = n
    index = i + 1

print(max)
print(index)

3
29
39
12
57
74
40
85
91
91
8


# 2577번

세 개의 자연수 A, B, C가 주어질 때 A × B × C를 계산한 결과에 0부터 9까지 각각의 숫자가 몇 번씩 쓰였는지를 구하는 프로그램을 작성하시오.

예를 들어 A = 150, B = 266, C = 427 이라면 A × B × C = 150 × 266 × 427 = 17037300 이 되고, 계산한 결과 17037300 에는 0이 3번, 1이 1번, 3이 2번, 7이 2번 쓰였다.

In [ ]:
a = int(input())
b = int(input())
c = int(input())

n = a * b * c
arr = [0 for _ in range(10)]

while (n != 0):
  arr[n%10] += 1
  n //= 10

for i in range(10):
  print(arr[i])



150
266
427
3
1
0
2
0
0
0
2
0
0


# 3052번

두 자연수 A와 B가 있을 때, A%B는 A를 B로 나눈 나머지 이다. 예를 들어, 7, 14, 27, 38을 3으로 나눈 나머지는 1, 2, 0, 2이다. 

수 10개를 입력받은 뒤, 이를 42로 나눈 나머지를 구한다. 그 다음 서로 다른 값이 몇 개 있는지 출력하는 프로그램을 작성하시오.

In [ ]:
arr = set([])
for _ in range(10):
  arr.add(int(input()) % 42)

print(len(arr))


42
84
252
420
840
126
42
84
420
126
1


# 1546번

세준이는 기말고사를 망쳤다. 세준이는 점수를 조작해서 집에 가져가기로 했다. 일단 세준이는 자기 점수 중에 최댓값을 골랐다. 이 값을 M이라고 한다. 그리고 나서 모든 점수를 점수/M*100으로 고쳤다.

예를 들어, 세준이의 최고점이 70이고, 수학점수가 50이었으면 수학점수는 50/70*100이 되어 71.43점이 된다.

세준이의 성적을 위의 방법대로 새로 계산했을 때, 새로운 평균을 구하는 프로그램을 작성하시오.

In [ ]:
n = int(input())

scores = list(map(int, input().split()))
max = scores[0]

for score in scores:
  if (score > max):
    max = score

average = 0
for i in range(n):
  average += (scores[i] / max * 100)

average /= n
print(average)



3
10 20 30
66.66666666666667


# 8958번

"OOXXOXXOOO"와 같은 OX퀴즈의 결과가 있다. O는 문제를 맞은 것이고, X는 문제를 틀린 것이다. 문제를 맞은 경우 그 문제의 점수는 그 문제까지 연속된 O의 개수가 된다. 예를 들어, 10번 문제의 점수는 3이 된다.

"OOXXOXXOOO"의 점수는 1+2+0+0+1+0+0+1+2+3 = 10점이다.

OX퀴즈의 결과가 주어졌을 때, 점수를 구하는 프로그램을 작성하시오.

In [ ]:
n = int(input())
scores = [0 for _ in range(n)]

for _ in range(n):
  line = input().split('X')
  score = 0
  for q in line:
    for i in range(len(q)):
      score += (i + 1)
  print(score)

5
OOXXOXXOOO
10
OOXXOOXXOO
9
OXOXOXOXOXOXOX
7
OOOOOOOOOO
55
OOOOXOOOOXOOOOX
30


# 4344번

대학생 새내기들의 90%는 자신이 반에서 평균은 넘는다고 생각한다. 당신은 그들에게 슬픈 진실을 알려줘야 한다.

In [ ]:
num_test_case = int(input())

for _ in range(num_test_case):
  line = list(map(int, input().split()))
  num_students = line[0]
  average = 0
  for i in range(1, len(line)):
    average += line[i]

  average /= num_students
  ratio = 0
  for i in range(1, len(line)):
    if (line[i] > average):
      ratio += 1
  ratio /= num_students
  ratio *= 100
  print(f'{ratio:.3f}%')


1
3 70 90 80
33.333%


# 15596번

정수 n개가 주어졌을 때, n개의 합을 구하는 함수를 작성하시오.

In [ ]:
def solve(a: list) -> int:
  sum = 0
  for num in a:
    sum += num
  return sum

# 4673번

셀프 넘버는 1949년 인도 수학자 D.R. Kaprekar가 이름 붙였다. 양의 정수 n에 대해서 d(n)을 n과 n의 각 자리수를 더하는 함수라고 정의하자. 예를 들어, d(75) = 75+7+5 = 87이다.

양의 정수 n이 주어졌을 때, 이 수를 시작해서 n, d(n), d(d(n)), d(d(d(n))), ...과 같은 무한 수열을 만들 수 있다. 

예를 들어, 33으로 시작한다면 다음 수는 33 + 3 + 3 = 39이고, 그 다음 수는 39 + 3 + 9 = 51, 다음 수는 51 + 5 + 1 = 57이다. 이런식으로 다음과 같은 수열을 만들 수 있다.

33, 39, 51, 57, 69, 84, 96, 111, 114, 120, 123, 129, 141, ...

n을 d(n)의 생성자라고 한다. 위의 수열에서 33은 39의 생성자이고, 39는 51의 생성자, 51은 57의 생성자이다. 생성자가 한 개보다 많은 경우도 있다. 예를 들어, 101은 생성자가 2개(91과 100) 있다. 

생성자가 없는 숫자를 셀프 넘버라고 한다. 100보다 작은 셀프 넘버는 총 13개가 있다. 1, 3, 5, 7, 9, 20, 31, 42, 53, 64, 75, 86, 97

10000보다 작거나 같은 셀프 넘버를 한 줄에 하나씩 출력하는 프로그램을 작성하시오.

In [ ]:
eratos = set(range(1, 10001))
self_num = []

def d(n):
  self_num = n
  for i in str(n):
    self_num += int(i)
  return self_num

while (len(eratos) != 0):
  n = eratos.pop()
  self_num.append(n)
  while (n <= 10000):
    n = d(n)
    eratos.discard(n)

for num in self_num:
  print(num)

[1, 3, 5, 7, 9, 20, 31, 42, 53, 64, 75, 86, 97, 108, 110, 121, 132, 143, 154, 165, 176, 187, 198, 209, 211, 222, 233, 244, 255, 266, 277, 288, 299, 310, 312, 323, 334, 345, 356, 367, 378, 389, 400, 411, 413, 424, 435, 446, 457, 468, 479, 490, 501, 512, 514, 525, 536, 547, 558, 569, 580, 591, 602, 613, 615, 626, 637, 648, 659, 670, 681, 692, 703, 714, 716, 727, 738, 749, 760, 771, 782, 793, 804, 815, 817, 828, 839, 850, 861, 872, 883, 894, 905, 916, 918, 929, 940, 951, 962, 973, 984, 995, 1006, 1021, 1032, 1043, 1054, 1065, 1076, 1087, 1098, 1109, 1111, 1122, 1133, 1144, 1155, 1166, 1177, 1188, 1199, 1210, 1212, 1223, 1234, 1245, 1256, 1267, 1278, 1289, 1300, 1311, 1313, 1324, 1335, 1346, 1357, 1368, 1379, 1390, 1401, 1412, 1414, 1425, 1436, 1447, 1458, 1469, 1480, 1491, 1502, 1513, 1515, 1526, 1537, 1548, 1559, 1570, 1581, 1592, 1603, 1614, 1616, 1627, 1638, 1649, 1660, 1671, 1682, 1693, 1704, 1715, 1717, 1728, 1739, 1750, 1761, 1772, 1783, 1794, 1805, 1816, 1818, 1829, 1840, 1851, 186

# 1065번

어떤 양의 정수 X의 각 자리가 등차수열을 이룬다면, 그 수를 한수라고 한다. 등차수열은 연속된 두 개의 수의 차이가 일정한 수열을 말한다. N이 주어졌을 때, 1보다 크거나 같고, N보다 작거나 같은 한수의 개수를 출력하는 프로그램을 작성하시오. 

In [ ]:
n = int(input())

def is_hansu(n):
  if (n < 10):
    return True

  b_ele = 10
  a_ele = 10
  dif = 10

  for digit in str(n):
    if (b_ele == 10):
      b_ele = int(digit)
    elif (a_ele == 10):
      a_ele = int(digit)
      dif = a_ele - b_ele
    else:
      b_ele = a_ele
      a_ele = int(digit)
      if (a_ele - b_ele != dif):
        return False

  return True

num_hansu = 0
for i in range(1, n+1):
  if (is_hansu(i)):
    num_hansu += 1

print(num_hansu)



210
105


# 11654번

알파벳 소문자, 대문자, 숫자 0-9중 하나가 주어졌을 때, 주어진 글자의 아스키 코드값을 출력하는 프로그램을 작성하시오.

In [ ]:
print(ord(input()))

A
65


# 11720번

N개의 숫자가 공백 없이 쓰여있다. 이 숫자를 모두 합해서 출력하는 프로그램을 작성하시오.

In [ ]:
n = int(input())
line = input()

sum = 0
for num in line:
  sum += int(num)
  
print(sum)

5
54321
15


# 10809번

알파벳 소문자로만 이루어진 단어 S가 주어진다. 각각의 알파벳에 대해서, 단어에 포함되어 있는 경우에는 처음 등장하는 위치를, 포함되어 있지 않은 경우에는 -1을 출력하는 프로그램을 작성하시오.

In [ ]:
def find(word, c):
  for i, l in enumerate(word):
    if (c == l):
      return i
  
  return -1

word = input()
characters = 'abcdefghijklmnopqrstuvwxyz'
result = ''

for c in characters:
  result += (str(find(word, c)) + ' ')

print(result[:len(result) - 1])

baekjoon
1 0 -1 -1 2 -1 -1 -1 -1 4 3 -1 -1 7 5 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1


# 2675번

문자열 S를 입력받은 후에, 각 문자를 R번 반복해 새 문자열 P를 만든 후 출력하는 프로그램을 작성하시오. 즉, 첫 번째 문자를 R번 반복하고, 두 번째 문자를 R번 반복하는 식으로 P를 만들면 된다. S에는 QR Code "alphanumeric" 문자만 들어있다.

QR Code "alphanumeric" 문자는 0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ\$%*+-./: 이다.

In [ ]:
num_case = int(input())
for _ in range(num_case):
  [iter, word] = input().split()
  iter = int(iter)
  new_word = ''
  for c in word:
    new_word += (c * iter)
  print(new_word)


2
3 ABC
AAABBBCCC
5 /HTP
/////HHHHHTTTTTPPPPP


# 1157번

알파벳 대소문자로 된 단어가 주어지면, 이 단어에서 가장 많이 사용된 알파벳이 무엇인지 알아내는 프로그램을 작성하시오. 단, 대문자와 소문자를 구분하지 않는다.

In [ ]:
word = input().upper()
char_dict = {}
for char in word:
  if char_dict.get(char) == None:
    char_dict[char] = 1
  else:
    char_dict[char] += 1

max = 0
for key, value in char_dict.items():
  if (value > max):
    max_key = key
    max = value

flag = 0
for key, value in char_dict.items():
  if (key != max_key and max == value):
    flag = 1
    break

if (flag):
  print('?')
else:
  print(max_key)

baaa
A


# 1152번

영어 대소문자와 공백으로 이루어진 문자열이 주어진다. 이 문자열에는 몇 개의 단어가 있을까? 이를 구하는 프로그램을 작성하시오. 단, 한 단어가 여러 번 등장하면 등장한 횟수만큼 모두 세어야 한다.

In [ ]:
sentence = input().split()
print(len(sentence))

tasad qwe qasdas qweqwe asd
5


# 2908번

상근이의 동생 상수는 수학을 정말 못한다. 상수는 숫자를 읽는데 문제가 있다. 이렇게 수학을 못하는 상수를 위해서 상근이는 수의 크기를 비교하는 문제를 내주었다. 상근이는 세 자리 수 두 개를 칠판에 써주었다. 그 다음에 크기가 큰 수를 말해보라고 했다.

상수는 수를 다른 사람과 다르게 거꾸로 읽는다. 예를 들어, 734와 893을 칠판에 적었다면, 상수는 이 수를 437과 398로 읽는다. 따라서, 상수는 두 수중 큰 수인 437을 큰 수라고 말할 것이다.

두 수가 주어졌을 때, 상수의 대답을 출력하는 프로그램을 작성하시오.

In [ ]:
[a, b] = input().split()
r_a, r_b = '', ''

for i in a[::-1]:
  r_a += i

for i in b[::-1]:
  r_b += i

r_a = int(r_a)
r_b = int(r_b)

if (r_a > r_b):
  print(r_a)
else:
  print(r_b)


734 893
437


# 5622

전화를 걸고 싶은 번호가 있다면, 숫자를 하나를 누른 다음에 금속 핀이 있는 곳 까지 시계방향으로 돌려야 한다. 숫자를 하나 누르면 다이얼이 처음 위치로 돌아가고, 다음 숫자를 누르려면 다이얼을 처음 위치에서 다시 돌려야 한다.

숫자 1을 걸려면 총 2초가 필요하다. 1보다 큰 수를 거는데 걸리는 시간은 이보다 더 걸리며, 한 칸 옆에 있는 숫자를 걸기 위해선 1초씩 더 걸린다.

상근이의 할머니는 전화 번호를 각 숫자에 해당하는 문자로 외운다. 즉, 어떤 단어를 걸 때, 각 알파벳에 해당하는 숫자를 걸면 된다. 예를 들어, UNUCIC는 868242와 같다.

할머니가 외운 단어가 주어졌을 때, 이 전화를 걸기 위해서 필요한 최소 시간을 구하는 프로그램을 작성하시오.

In [ ]:
chars = ['ABC', 'DEF', 'GHI', 'JKL', 'MNO', 'PQRS', 'TUV', 'WXYZ']
char_dict = {}
dial_dict = {}

for i, char in enumerate(chars):
  for c in char:
    char_dict[c] = i + 2

for i in range(1, 11):
  dial_dict[i%10] = i+1

nums = input()
sum = 0 
for num in nums:
  sum += dial_dict[char_dict[num]]

print(sum)


UNUCIC
36


# 2941번

예전에는 운영체제에서 크로아티아 알파벳을 입력할 수가 없었다. 따라서, 다음과 같이 크로아티아 알파벳을 변경해서 입력했다.

크로아티아 알파벳	변경
č	c=/
ć	c-/
dž	dz=/
đ	d-/
lj	lj/
nj	nj/
š	s=/
ž	z=/
예를 들어, ljes=njak은 크로아티아 알파벳 6개(lj, e, š, nj, a, k)로 이루어져 있다. 단어가 주어졌을 때, 몇 개의 크로아티아 알파벳으로 이루어져 있는지 출력한다.

dž는 무조건 하나의 알파벳으로 쓰이고, d와 ž가 분리된 것으로 보지 않는다. lj와 nj도 마찬가지이다. 위 목록에 없는 알파벳은 한 글자씩 센다

In [ ]:
croatia = ['c=', 'c-', 'dz=', 'd-', 'lj', 'nj', 's=', 'z=']
words = input()
for word in croatia:
  words = words.replace(word, '*')

print(len(words))

ljes=njak
6


# 1316번

그룹 단어란 단어에 존재하는 모든 문자에 대해서, 각 문자가 연속해서 나타나는 경우만을 말한다. 예를 들면, ccazzzzbb는 c, a, z, b가 모두 연속해서 나타나고, kin도 k, i, n이 연속해서 나타나기 때문에 그룹 단어이지만, aabbbccb는 b가 떨어져서 나타나기 때문에 그룹 단어가 아니다.

단어 N개를 입력으로 받아 그룹 단어의 개수를 출력하는 프로그램을 작성하시오.

In [ ]:
n = int(input())
words = []

num_group_word = n
for _ in range(n):
  words.append(input())

for word in words:
  char_set = set()
  prev_c = ''
  for c in word:
    if (prev_c == c):
      pass
    elif (c in char_set):
      num_group_word -= 1
      break
    else:
      char_set.add(c)
      prev_c = c
print(num_group_word)


4
aba
abab
abcabc
a
1


# 1712번

월드전자는 노트북을 제조하고 판매하는 회사이다. 노트북 판매 대수에 상관없이 매년 임대료, 재산세, 보험료, 급여 등 A만원의 고정 비용이 들며, 한 대의 노트북을 생산하는 데에는 재료비와 인건비 등 총 B만원의 가변 비용이 든다고 한다.

예를 들어 A=1,000, B=70이라고 하자. 이 경우 노트북을 한 대 생산하는 데는 총 1,070만원이 들며, 열 대 생산하는 데는 총 1,700만원이 든다.

노트북 가격이 C만원으로 책정되었다고 한다. 일반적으로 생산 대수를 늘려 가다 보면 어느 순간 총 수입(판매비용)이 총 비용(=고정비용+가변비용)보다 많아지게 된다. 최초로 총 수입이 총 비용보다 많아져 이익이 발생하는 지점을 손익분기점(BREAK-EVEN POINT)이라고 한다.

A, B, C가 주어졌을 때, 손익분기점을 구하는 프로그램을 작성하시오.

In [ ]:
[fixed_cost, cost, price] = [int(i) for i in input().split()]
margin = price - cost

if (margin <= 0):
  print(-1)
else:
  print(fixed_cost // margin + 1)

3 2 1
-1
-2


# 2292번

위의 그림과 같이 육각형으로 이루어진 벌집이 있다. 그림에서 보는 바와 같이 중앙의 방 1부터 시작해서 이웃하는 방에 돌아가면서 1씩 증가하는 번호를 주소로 매길 수 있다. 숫자 N이 주어졌을 때, 벌집의 중앙 1에서 N번 방까지 최소 개수의 방을 지나서 갈 때 몇 개의 방을 지나가는지(시작과 끝을 포함하여)를 계산하는 프로그램을 작성하시오. 예를 들면, 13까지는 3개, 58까지는 5개를 지난다.

In [ ]:
n = int(input())
distance = 1
while (n > 1 + 3 * distance * (distance - 1)):
  distance += 1
  
print(distance)

38
5


# 1193번

이와 같이 나열된 분수들을 1/1 → 1/2 → 2/1 → 3/1 → 2/2 → … 과 같은 지그재그 순서로 차례대로 1번, 2번, 3번, 4번, 5번, … 분수라고 하자.

X가 주어졌을 때, X번째 분수를 구하는 프로그램을 작성하시오.

In [ ]:
n = int(input())

order = 1
while(n > order * (order + 1) / 2):
  order += 1

if (order % 2 == 0):
  over = n - order * (order - 1) // 2
  under = order * (order + 1) // 2 - n + 1
else:
  over = order * (order + 1) // 2 - n + 1
  under = n - order * (order - 1) // 2
  

print(f'{over}/{under}')


14
2/4


# 2869번

땅 위에 달팽이가 있다. 이 달팽이는 높이가 V미터인 나무 막대를 올라갈 것이다.

달팽이는 낮에 A미터 올라갈 수 있다. 하지만, 밤에 잠을 자는 동안 B미터 미끄러진다. 또, 정상에 올라간 후에는 미끄러지지 않는다.

달팽이가 나무 막대를 모두 올라가려면, 며칠이 걸리는지 구하는 프로그램을 작성하시오.

In [ ]:
import math

[up, down, distance] = [int(i) for i in input().split()]
print(math.ceil((distance - up) / (up - down)) + 1)

5 1 6
2


# 10250번

방 번호는 YXX 나 YYXX 형태인데 여기서 Y 나 YY 는 층 수를 나타내고 XX 는 엘리베이터에서부터 세었을 때의 번호를 나타낸다. 즉, 그림 1 에서 빗금으로 표시한 방은 305 호가 된다.

손님은 엘리베이터를 타고 이동하는 거리는 신경 쓰지 않는다. 다만 걷는 거리가 같을 때에는 아래층의 방을 더 선호한다. 예를 들면 102 호 방보다는 301 호 방을 더 선호하는데, 102 호는 거리 2 만큼 걸어야 하지만 301 호는 거리 1 만큼만 걸으면 되기 때문이다. 같은 이유로 102 호보다 2101 호를 더 선호한다.

여러분이 작성할 프로그램은 초기에 모든 방이 비어있다고 가정하에 이 정책에 따라 N 번째로 도착한 손님에게 배정될 방 번호를 계산하는 프로그램이다. 첫 번째 손님은 101 호, 두 번째 손님은 201 호 등과 같이 배정한다. 그림 1 의 경우를 예로 들면, H = 6이므로 10 번째 손님은 402 호에 배정해야 한다.

In [ ]:
import math

num_test_case = int(input())

for _ in range(num_test_case):
  [height, width, n] = [int(i) for i in input().split()]
  x = math.ceil(n / height)
  y = n % height
  if (y == 0):
    y = height
  print(f'{y}{x//10}{x%10}')
  

1
5 2 3
301


# 2775번

평소 반상회에 참석하는 것을 좋아하는 주희는 이번 기회에 부녀회장이 되고 싶어 각 층의 사람들을 불러 모아 반상회를 주최하려고 한다.

이 아파트에 거주를 하려면 조건이 있는데, “a층의 b호에 살려면 자신의 아래(a-1)층의 1호부터 b호까지 사람들의 수의 합만큼 사람들을 데려와 살아야 한다” 는 계약 조항을 꼭 지키고 들어와야 한다.

아파트에 비어있는 집은 없고 모든 거주민들이 이 계약 조건을 지키고 왔다고 가정했을 때, 주어지는 양의 정수 k와 n에 대해 k층에 n호에는 몇 명이 살고 있는지 출력하라. 단, 아파트에는 0층부터 있고 각층에는 1호부터 있으며, 0층의 i호에는 i명이 산다.

In [ ]:
num_test_case = int(input())
for _ in range(num_test_case):
  floor = int(input())
  room = int(input())

  num_people = [i for i in range(1, room + 1)]
  for _ in range(floor):
    for i in range(room)[::-1]:
      num_people[i] = sum(num_people[:i+1])

  print(num_people[room-1])

2
1
3
[1, 3, 6]
2
3
[1, 4, 10]


# 2839번

상근이는 요즘 설탕공장에서 설탕을 배달하고 있다. 상근이는 지금 사탕가게에 설탕을 정확하게 N킬로그램을 배달해야 한다. 설탕공장에서 만드는 설탕은 봉지에 담겨져 있다. 봉지는 3킬로그램 봉지와 5킬로그램 봉지가 있다.

상근이는 귀찮기 때문에, 최대한 적은 봉지를 들고 가려고 한다. 예를 들어, 18킬로그램 설탕을 배달해야 할 때, 3킬로그램 봉지 6개를 가져가도 되지만, 5킬로그램 3개와 3킬로그램 1개를 배달하면, 더 적은 개수의 봉지를 배달할 수 있다.

상근이가 설탕을 정확하게 N킬로그램 배달해야 할 때, 봉지 몇 개를 가져가면 되는지 그 수를 구하는 프로그램을 작성하시오.

In [ ]:
kg_n = int(input())

kg_5 = 0
kg_3 = 0

available_kg = [i + j for i in [0, 5, 10] for j in [0, 3, 6, 9, 12]]

while (kg_n > 0):

  if (kg_n in available_kg):
    for i in range(3):
      for j in range(5):
        if (available_kg[5*i +j] == kg_n):
          kg_5 += i
          kg_3 += j
    break
  else:
    kg_n -= 15
    kg_5 += 3

if (kg_n < 0):
  print(-1)
else:
  print(kg_5 + kg_3)

38
8


# 10757번

두 정수 A와 B를 입력받은 다음, A+B를 출력하는 프로그램을 작성하시오.

In [ ]:
[A, B] = [int(i) for i in input().split()]
print(A + B)

9223372036854775807 9223372036854775808
18446744073709551615


# 1978번

주어진 수 N개 중에서 소수가 몇 개인지 찾아서 출력하는 프로그램을 작성하시오.

In [ ]:
eratos = [i for i in range(1001)]
prime_nums = set([])

eratos[0] = False
eratos[1] = False

for prime in eratos:
  if (eratos[prime]):
    prime_nums.add(prime)
    for i in eratos[2*prime::prime]:
      eratos[i] = False

n = int(input())
nums = list(map(int, input().split()))

cnt = 0
for num in nums:
  if (num in prime_nums):
    cnt += 1

print(cnt)


4
1 3 5 7
3


# 2581번

자연수 M과 N이 주어질 때 M이상 N이하의 자연수 중 소수인 것을 모두 골라 이들 소수의 합과 최솟값을 찾는 프로그램을 작성하시오.

예를 들어 M=60, N=100인 경우 60이상 100이하의 자연수 중 소수는 61, 67, 71, 73, 79, 83, 89, 97 총 8개가 있으므로, 이들 소수의 합은 620이고, 최솟값은 61이 된다.

In [ ]:
eratos = [i for i in range(10001)]
prime_nums = set([])

eratos[0] = False
eratos[1] = False

for prime in eratos:
  if (eratos[prime]):
    prime_nums.add(prime)
    for i in eratos[2*prime::prime]:
      eratos[i] = False

M = int(input())
N = int(input())

primes = []
for i in range(M, N+1):
  if (i in prime_nums):
    primes.append(i)

if (len(primes) == 0):
  print(-1)
else:
  print(sum(primes))
  print(primes[0])

60
100
620
61


#11635번

정수 N이 주어졌을 때, 소인수분해하는 프로그램을 작성하시오.

In [ ]:
n = int(input())
if (n == 1):
  pass
else:
  factors = []
  for p in range(2, n+1):
    if (p > n):
      break
    while(n % p == 0):
      n //= p
      factors.append(p)
  for i in factors:
    print(i)


9991
97
103


# 1929번

M이상 N이하의 소수를 모두 출력하는 프로그램을 작성하시오.

In [ ]:
[M, N] = [int(i) for i in input().split()]

eratos = [i for i in range(N+1)]
prime_nums = []

eratos[0] = False
eratos[1] = False

for prime in eratos:
  if (eratos[prime]):
    if (prime >= M):
      prime_nums.append(prime)
    for i in eratos[2*prime::prime]:
      eratos[i] = False

for i in prime_nums:
  print(i)

3 16
3
5
7
11
13


#4948번

베르트랑 공준은 임의의 자연수 n에 대하여, n보다 크고, 2n보다 작거나 같은 소수는 적어도 하나 존재한다는 내용을 담고 있다.

이 명제는 조제프 베르트랑이 1845년에 추측했고, 파프누티 체비쇼프가 1850년에 증명했다.

예를 들어, 10보다 크고, 20보다 작거나 같은 소수는 4개가 있다. (11, 13, 17, 19) 또, 14보다 크고, 28보다 작거나 같은 소수는 3개가 있다. (17,19, 23)

자연수 n이 주어졌을 때, n보다 크고, 2n보다 작거나 같은 소수의 개수를 구하는 프로그램을 작성하시오. 

In [ ]:
N = int(input())
while(N != 0):
  eratos = [i for i in range(2*N+1)]
  prime_nums = 0

  eratos[0] = False
  eratos[1] = False

  for prime in eratos:
    if (eratos[prime]):
      if (prime > N):
        prime_nums += 1
      for i in eratos[2*prime::prime]:
        eratos[i] = False
  print(prime_nums)
  N = int(input())

1
1
10
4
13
3
100
21
1000
135
10000
1033
100000
8392
0


# 9020번

1보다 큰 자연수 중에서  1과 자기 자신을 제외한 약수가 없는 자연수를 소수라고 한다. 예를 들어, 5는 1과 5를 제외한 약수가 없기 때문에 소수이다. 하지만, 6은 6 = 2 × 3 이기 때문에 소수가 아니다.

골드바흐의 추측은 유명한 정수론의 미해결 문제로, 2보다 큰 모든 짝수는 두 소수의 합으로 나타낼 수 있다는 것이다. 이러한 수를 골드바흐 수라고 한다. 또, 짝수를 두 소수의 합으로 나타내는 표현을 그 수의 골드바흐 파티션이라고 한다. 예를 들면, 4 = 2 + 2, 6 = 3 + 3, 8 = 3 + 5, 10 = 5 + 5, 12 = 5 + 7, 14 = 3 + 11, 14 = 7 + 7이다. 10000보다 작거나 같은 모든 짝수 n에 대한 골드바흐 파티션은 존재한다.

2보다 큰 짝수 n이 주어졌을 때, n의 골드바흐 파티션을 출력하는 프로그램을 작성하시오. 만약 가능한 n의 골드바흐 파티션이 여러 가지인 경우에는 두 소수의 차이가 가장 작은 것을 출력한다

In [ ]:
def get_primes(n):
  sieve = [False, False] + [True] * (n - 1)
  primes = []
  for i in range(int(n ** 0.5) + 1):
    if (sieve[i]):
      sieve[2*i::i] = [False] * (n // i - 1)

  return [i for i, flag in enumerate(sieve) if flag]

num_test_case = int(input())

for _ in range(num_test_case):
  n = int(input())
  if (n % 2 != 0):
    continue
  primes = get_primes(n)
  prime_set = set(primes)
  for i in primes:
    if (i >= n // 2 + 1):
      break
    if (n - i in prime_set):
      goldbach_num = [i, n - i]
      
  print(goldbach_num[0], goldbach_num[1])


3
8888
4441 4447
10000
4919 5081
1000000
499943 500057


# 10872번

0보다 크거나 같은 정수 N이 주어진다. 이때, N!을 출력하는 프로그램을 작성하시오.

In [ ]:
def factorial(n):
  if (n <= 0):
    return 1
  return n * factorial(n - 1)

n = int(input())
print(factorial(n))

10
3628800


# 10870번

피보나치 수는 0과 1로 시작한다. 0번째 피보나치 수는 0이고, 1번째 피보나치 수는 1이다. 그 다음 2번째 부터는 바로 앞 두 피보나치 수의 합이 된다.

이를 식으로 써보면 Fn = Fn-1 + Fn-2 (n ≥ 2)가 된다.

n=17일때 까지 피보나치 수를 써보면 다음과 같다.

0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597

n이 주어졌을 때, n번째 피보나치 수를 구하는 프로그램을 작성하시오.

In [ ]:
def fibonacci(n):
  if (n == 0):
    return 0
  elif (n == 1):
    return 1
  else:
    return fibonacci(n - 1) + fibonacci(n - 2)

n = int(input())
print(fibonacci(n))

17
1597


# 17478번

평소에 질문을 잘 받아주기로 유명한 중앙대학교의 JH 교수님은 학생들로부터 재귀함수가 무엇인지에 대하여 많은 질문을 받아왔다.

매번 질문을 잘 받아주셨던 JH 교수님이지만 그는 중앙대학교가 자신과 맞는가에 대한 고민을 항상 해왔다.

중앙대학교와 자신의 길이 맞지 않다고 생각한 JH 교수님은 결국 중앙대학교를 떠나기로 결정하였다.

떠나기 전까지도 제자들을 생각하셨던 JH 교수님은 재귀함수가 무엇인지 물어보는 학생들을 위한 작은 선물로 자동 응답 챗봇을 준비하기로 했다.

JH 교수님이 만들 챗봇의 응답을 출력하는 프로그램을 만들어보자.

In [ ]:
def what_is_recursion(n, iter = 0):
  if (iter == 0):
    print('어느 한 컴퓨터공학과 학생이 유명한 교수님을 찾아가 물었다.')
    
  
  print('____' * iter + '"재귀함수가 뭔가요?"')
  if (iter == n):
    print('____' * iter + '"재귀함수는 자기 자신을 호출하는 함수라네"')
    print('____' * iter + '라고 답변하였지.')
    return
  else:
    print('____' * iter + '"잘 들어보게. 옛날옛날 한 산 꼭대기에 이세상 모든 지식을 통달한 선인이 있었어.')
    print('____' * iter + '마을 사람들은 모두 그 선인에게 수많은 질문을 했고, 모두 지혜롭게 대답해 주었지.')
    print('____' * iter + '그의 답은 대부분 옳았다고 하네. 그런데 어느 날, 그 선인에게 한 선비가 찾아와서 물었어."')
    what_is_recursion(n, iter + 1)
  
  print('____' * iter + '라고 답변하였지.')

n = int(input())
what_is_recursion(n)

2
어느 한 컴퓨터공학과 학생이 유명한 교수님을 찾아가 물었다.
"재귀함수가 뭔가요?"
"잘 들어보게. 옛날옛날 한 산 꼭대기에 이세상 모든 지식을 통달한 선인이 있었어.
마을 사람들은 모두 그 선인에게 수많은 질문을 했고, 모두 지혜롭게 대답해 주었지.
그의 답은 대부분 옳았다고 하네. 그런데 어느 날, 그 선인에게 한 선비가 찾아와서 물었어."
----"재귀함수가 뭔가요?"
----"잘 들어보게. 옛날옛날 한 산 꼭대기에 이세상 모든 지식을 통달한 선인이 있었어.
----마을 사람들은 모두 그 선인에게 수많은 질문을 했고, 모두 지혜롭게 대답해 주었지.
----그의 답은 대부분 옳았다고 하네. 그런데 어느 날, 그 선인에게 한 선비가 찾아와서 물었어."
--------"재귀함수가 뭔가요?"
--------"재귀함수는 자기 자신을 호출하는 함수라네"
--------라고 답변하였지.
----라고 답변하였지.
라고 답변하였지.


# 2447번

재귀적인 패턴으로 별을 찍어 보자. N이 3의 거듭제곱(3, 9, 27, ...)이라고 할 때, 크기 N의 패턴은 N×N 정사각형 모양이다.

크기 3의 패턴은 가운데에 공백이 있고, 가운데를 제외한 모든 칸에 별이 하나씩 있는 패턴이다.

```
***
* *
***
```

N이 3보다 클 경우, 크기 N의 패턴은 공백으로 채워진 가운데의 (N/3)×(N/3) 정사각형을 크기 N/3의 패턴으로 둘러싼 형태이다. 예를 들어 크기 27의 패턴은 예제 출력 1과 같다.

In [ ]:
def print_star(n, arr = ['*']):
  arr *= 3
  for i in range(len(arr)):
    if(len(arr) // 3 <= i and i < len(arr) * 2 // 3):
      arr[i] = arr[i] + ' ' * (len(arr) // 3) + arr[i]
    else:
      arr[i] *= 3
  if (len(arr) == n):
    for line in arr:
      print(line)
  else:
    print_star(n, arr)

n = int(input())
print_star(n)



27
***************************
* ** ** ** ** ** ** ** ** *
***************************
***   ******   ******   ***
* *   * ** *   * ** *   * *
***   ******   ******   ***
***************************
* ** ** ** ** ** ** ** ** *
***************************
*********         *********
* ** ** *         * ** ** *
*********         *********
***   ***         ***   ***
* *   * *         * *   * *
***   ***         ***   ***
*********         *********
* ** ** *         * ** ** *
*********         *********
***************************
* ** ** ** ** ** ** ** ** *
***************************
***   ******   ******   ***
* *   * ** *   * ** *   * *
***   ******   ******   ***
***************************
* ** ** ** ** ** ** ** ** *
***************************


# 11729번

세 개의 장대가 있고 첫 번째 장대에는 반경이 서로 다른 n개의 원판이 쌓여 있다. 각 원판은 반경이 큰 순서대로 쌓여있다. 이제 수도승들이 다음 규칙에 따라 첫 번째 장대에서 세 번째 장대로 옮기려 한다.

한 번에 한 개의 원판만을 다른 탑으로 옮길 수 있다.
쌓아 놓은 원판은 항상 위의 것이 아래의 것보다 작아야 한다.
이 작업을 수행하는데 필요한 이동 순서를 출력하는 프로그램을 작성하라. 단, 이동 횟수는 최소가 되어야 한다.

In [ ]:
def hanoi(n, src = 1, dst = 3, stp = 2, cnt = [0]):
  if (n == 1):
    cnt.append(str(src) + ' ' + str(dst))
    cnt[0] += 1
    return
  hanoi(n - 1, src, stp, dst, cnt)
  cnt.append(str(src) + ' ' + str(dst))
  cnt[0] += 1
  hanoi(n - 1, stp, dst, src, cnt)
  

n = int(input())
cnt = [0]
hanoi(n, cnt=cnt)
for line in cnt:
  print(line)


3
7
1 3
1 2
3 2
1 3
2 1
2 3
1 3


In [ ]:
def rec(arr, n):
  if (n == 0):
    return
  arr.append(n)
  rec(arr, n - 1)

arr = []
rec(arr, 10)
print(arr)

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]


# 2798번

카지노에서 제일 인기 있는 게임 블랙잭의 규칙은 상당히 쉽다. 카드의 합이 21을 넘지 않는 한도 내에서, 카드의 합을 최대한 크게 만드는 게임이다. 블랙잭은 카지노마다 다양한 규정이 있다.

한국 최고의 블랙잭 고수 김정인은 새로운 블랙잭 규칙을 만들어 상근, 창영이와 게임하려고 한다.

김정인 버전의 블랙잭에서 각 카드에는 양의 정수가 쓰여 있다. 그 다음, 딜러는 N장의 카드를 모두 숫자가 보이도록 바닥에 놓는다. 그런 후에 딜러는 숫자 M을 크게 외친다.

이제 플레이어는 제한된 시간 안에 N장의 카드 중에서 3장의 카드를 골라야 한다. 블랙잭 변형 게임이기 때문에, 플레이어가 고른 카드의 합은 M을 넘지 않으면서 M과 최대한 가깝게 만들어야 한다.

N장의 카드에 써져 있는 숫자가 주어졌을 때, M을 넘지 않으면서 M에 최대한 가까운 카드 3장의 합을 구해 출력하시오.

In [ ]:
def black_jack(M, N_nums):
  max_sum = 0
  for i, first_card in enumerate(N_nums):
    for j, second_card in enumerate(N_nums[i+1:]):
      for third_card in N_nums[i+j+2:]:
        sum_card = first_card + second_card + third_card
        if (sum_card > max_sum and sum_card <= M):
          max_sum = sum_card

  return max_sum

[N, M] = [int(i) for i in input().split()]
N_nums = list(map(int, input().split()))

sum_card = black_jack(M, N_nums)
print(sum_card)

10 500
93 181 245 214 315 36 185 138 216 295
497


# 2231번

어떤 자연수 N이 있을 때, 그 자연수 N의 분해합은 N과 N을 이루는 각 자리수의 합을 의미한다. 어떤 자연수 M의 분해합이 N인 경우, M을 N의 생성자라 한다. 예를 들어, 245의 분해합은 256(=245+2+4+5)이 된다. 따라서 245는 256의 생성자가 된다. 물론, 어떤 자연수의 경우에는 생성자가 없을 수도 있다. 반대로, 생성자가 여러 개인 자연수도 있을 수 있다.

자연수 N이 주어졌을 때, N의 가장 작은 생성자를 구해내는 프로그램을 작성하시오.

In [ ]:
def digitize(n):
  n_digits = []
  while n >= 10:
    n_digits.append(n % 10)
    n //= 10
  n_digits.append(n)
  return n_digits

def find_constructor(n):
  X = []
  for sum_digits in range(100):
    if (sum(digitize(n - sum_digits)) == sum_digits):
      X.append(n - sum_digits)
  if (X == []):
    return 0
  else:
    return min(X)
    
n = int(input())
print(find_constructor(n))

216
198


# 7568번

우리는 사람의 덩치를 키와 몸무게, 이 두 개의 값으로 표현하여 그 등수를 매겨보려고 한다. 어떤 사람의 몸무게가 x kg이고 키가 y cm라면 이 사람의 덩치는 (x, y)로 표시된다. 두 사람 A 와 B의 덩치가 각각 (x, y), (p, q)라고 할 때 x > p 그리고 y > q 이라면 우리는 A의 덩치가 B의 덩치보다 "더 크다"고 말한다. 예를 들어 어떤 A, B 두 사람의 덩치가 각각 (56, 177), (45, 165) 라고 한다면 A의 덩치가 B보다 큰 셈이 된다. 그런데 서로 다른 덩치끼리 크기를 정할 수 없는 경우도 있다. 예를 들어 두 사람 C와 D의 덩치가 각각 (45, 181), (55, 173)이라면 몸무게는 D가 C보다 더 무겁고, 키는 C가 더 크므로, "덩치"로만 볼 때 C와 D는 누구도 상대방보다 더 크다고 말할 수 없다.

N명의 집단에서 각 사람의 덩치 등수는 자신보다 더 "큰 덩치"의 사람의 수로 정해진다. 만일 자신보다 더 큰 덩치의 사람이 k명이라면 그 사람의 덩치 등수는 k+1이 된다. 이렇게 등수를 결정하면 같은 덩치 등수를 가진 사람은 여러 명도 가능하다. 아래는 5명으로 이루어진 집단에서 각 사람의 덩치와 그 등수가 표시된 표이다.

In [ ]:
def bigger(x, y):
  return x[0] > y[0] and x[1] > y[1]

n = int(input())
people = []
for _ in range(n):
  people.append([int(attr) for attr in input().split()])

for i, person in enumerate(people):
  person.append(0)
  for j, opponent in enumerate(people):
    if (i == j):
      continue
    if (bigger(opponent, person)):
      person[2] += 1

line = ''
for i in range(n):
  line += (str(people[i][2] + 1) + ' ')

print(line[:-1])
  




5
55 185
58 183
88 186
60 175
46 155
2 2 1 2 5


# 1018번

지민이는 자신의 저택에서 MN개의 단위 정사각형으로 나누어져 있는 M×N 크기의 보드를 찾았다. 어떤 정사각형은 검은색으로 칠해져 있고, 나머지는 흰색으로 칠해져 있다. 지민이는 이 보드를 잘라서 8×8 크기의 체스판으로 만들려고 한다.

체스판은 검은색과 흰색이 번갈아서 칠해져 있어야 한다. 구체적으로, 각 칸이 검은색과 흰색 중 하나로 색칠되어 있고, 변을 공유하는 두 개의 사각형은 다른 색으로 칠해져 있어야 한다. 따라서 이 정의를 따르면 체스판을 색칠하는 경우는 두 가지뿐이다. 하나는 맨 왼쪽 위 칸이 흰색인 경우, 하나는 검은색인 경우이다.

보드가 체스판처럼 칠해져 있다는 보장이 없어서, 지민이는 8×8 크기의 체스판으로 잘라낸 후에 몇 개의 정사각형을 다시 칠해야겠다고 생각했다. 당연히 8*8 크기는 아무데서나 골라도 된다. 지민이가 다시 칠해야 하는 정사각형의 최소 개수를 구하는 프로그램을 작성하시오.

In [ ]:
def num_gap(board):
  gap = 0
  origin_board = [
                  'WBWBWBWB',
                  'BWBWBWBW',
                  'WBWBWBWB',
                  'BWBWBWBW',
                  'WBWBWBWB',
                  'BWBWBWBW',
                  'WBWBWBWB',
                  'BWBWBWBW'
  ]

  for row in range(8):
    for col in range(8):
      if (board[row][col] != origin_board[row][col]):
        gap += 1

  min_gap = gap
  gap = 0
  for row in range(8):
    for col in range(8):
      if (board[-(row + 1)][col] != origin_board[row][col]):
        gap += 1

  return min(gap, min_gap)  

[height, width] = [int(i) for i in input().split()]
board = []
for row in range(height):
  board.append(input())

min_gap = 64

for row in range(height - 7):
  for col in range(width - 7):
    gap = num_gap([line[col:col+8] for line in board[row:row+8]])
    if (gap < min_gap):
      min_gap = gap

print(min_gap)

8 8
BBBBBBBB
BBBBBBBB
BBBBBBBB
BBBBBBBB
BBBBBBBB
BBBBBBBB
BBBBBBBB
BBBBBBBB
32


# 1436번

666은 종말을 나타내는 숫자라고 한다. 따라서, 많은 블록버스터 영화에서는 666이 들어간 제목을 많이 사용한다. 영화감독 숌은 세상의 종말 이라는 시리즈 영화의 감독이다. 조지 루카스는 스타워즈를 만들 때, 스타워즈 1, 스타워즈 2, 스타워즈 3, 스타워즈 4, 스타워즈 5, 스타워즈 6과 같이 이름을 지었고, 피터 잭슨은 반지의 제왕을 만들 때, 반지의 제왕 1, 반지의 제왕 2, 반지의 제왕 3과 같이 영화 제목을 지었다.

하지만 숌은 자신이 조지 루카스와 피터 잭슨을 뛰어넘는다는 것을 보여주기 위해서 영화 제목을 좀 다르게 만들기로 했다.

종말의 숫자란 어떤 수에 6이 적어도 3개이상 연속으로 들어가는 수를 말한다. 제일 작은 종말의 숫자는 666이고, 그 다음으로 큰 수는 1666, 2666, 3666, .... 과 같다.

따라서, 숌은 첫 번째 영화의 제목은 세상의 종말 666, 두 번째 영화의 제목은 세상의 종말 1666 이렇게 이름을 지을 것이다. 일반화해서 생각하면, N번째 영화의 제목은 세상의 종말 (N번째로 작은 종말의 숫자) 와 같다.

숌이 만든 N번째 영화의 제목에 들어간 숫자를 출력하는 프로그램을 작성하시오. 숌은 이 시리즈를 항상 차례대로 만들고, 다른 영화는 만들지 않는다.

In [ ]:
N = int(input())
end_nums = []
for num in range(3000000):
  if (str(num).find('666') != -1):
    end_nums.append(num)

print(end_nums[N - 1])

2342
666541


# 2750번

첫째 줄에 수의 개수 N(1 ≤ N ≤ 1,000)이 주어진다. 둘째 줄부터 N개의 줄에는 수 주어진다. 이 수는 절댓값이 1,000보다 작거나 같은 정수이다. 수는 중복되지 않는다.

In [ ]:
n = int(input())
nums = []
for _ in range(n):
  nums.append(int(input()))

nums.sort()
for num in nums:
  print(num)

5
5
2
3
4
1
1
2
3
4
5


# 2751번

첫째 줄에 수의 개수 N(1 ≤ N ≤ 1,000,000)이 주어진다. 둘째 줄부터 N개의 줄에는 수가 주어진다. 이 수는 절댓값이 1,000,000보다 작거나 같은 정수이다. 수는 중복되지 않는다.

In [ ]:
import sys

n = int(input())
nums = []
for _ in range(n):
  nums.append(int(sys.stdin.readline()))

nums.sort()
for num in nums:
  print(num)

# 10989번

N개의 수가 주어졌을 때, 이를 오름차순으로 정렬하는 프로그램을 작성하시오.

In [ ]:
class Heap:
  def __init__(self, capacity = 0):
    self.size = 0
    self.capacity = capacity +1
    self.heap = list(range(self.capacity))

  def insert(self, data):
    self.size += 1
    if (self.size < self.capacity):
      self.heap[self.size] = data

    else:
      self.heap.append(data)
      self.capacity += 1

    index = self.size
    while (index > 0 and self.heap[index // 2] > data):

      self.heap[index] = self.heap[index // 2]
      self.heap[index // 2] = data
      index //= 2


n = int(input())
h = Heap(n)
for _ in range(n):
  h.insert(int(input()))

for num in h.heap[1:]:
  print(num)

10
5
2
3
1
4
2
3
5
1
7
1
1
2
2
4
3
3
5
5
7


In [ ]:
import sys

n = int(input())
arr = [0] * 10000
for _ in range(n):
  arr[int(sys.stdin.readline()) - 1] += 1

for num, i in enumerate(arr):
  for j in range(i):
    print(num + 1)

10
5
2
3
1
4
2
3
5
1
7
1
1
2
2
3
3
4
5
5
7


# 2108번

수를 처리하는 것은 통계학에서 상당히 중요한 일이다. 통계학에서 N개의 수를 대표하는 기본 통계값에는 다음과 같은 것들이 있다. 단, N은 홀수라고 가정하자.

1. 산술평균 : N개의 수들의 합을 N으로 나눈 값
2. 중앙값 : N개의 수들을 증가하는 순서로 나열했을 경우 그 중앙에 위치하는 값
3. 최빈값 : N개의 수들 중 가장 많이 나타나는 값
4. 범위 : N개의 수들 중 최댓값과 최솟값의 차이

N개의 수가 주어졌을 때, 네 가지 기본 통계값을 구하는 프로그램을 작성하시오.

In [6]:
n = int(input())
n_arr = []

for _ in range(n):
  n_arr.append(int(input()))

n_arr.sort()

num_dict = {}
for num in n_arr:
  if num in num_dict:
    num_dict[num] += 1
  else:
    num_dict[num] = 1
  
mode = []
max_freq = max(num_dict.values())
for num, freq in num_dict.items():
  if freq == max_freq:
    mode.append(num)

if (len(mode) > 1):
  mode.sort()
  mode = mode[1]
else:
  mode = mode[0]

print(round(sum(n_arr) / n))
print(n_arr[n // 2])
print(mode)
print(n_arr[-1] - n_arr[0])

5
1
3
8
-2
2
2
2
1
10


# 1427번

배열을 정렬하는 것은 쉽다. 수가 주어지면, 그 수의 각 자리수를 내림차순으로 정렬해보자.

In [7]:
n = input()
n_digit = [int(digit) for digit in n]
n_digit.sort(reverse = True)

n_str = ''
for digit in n_digit:
  n_str += str(digit)

print(n_str)


6465123
6654321


# 11650번

2차원 평면 위의 점 N개가 주어진다. 좌표를 x좌표가 증가하는 순으로, x좌표가 같으면 y좌표가 증가하는 순서로 정렬한 다음 출력하는 프로그램을 작성하시오.

In [8]:
n = int(input())

coordinate = {}
x_arr = []
for _ in range(n):
  [x, y] = [int(i) for i in input().split()]
  if x in coordinate:
    coordinate[x].append(y)
  else:
    coordinate[x] = [y]
    x_arr.append(x)

x_arr.sort()

for x in x_arr:
  y_arr = coordinate[x]
  y_arr.sort()
  for y in y_arr:
    print(x, y)


5
3 4
1 1
1 -1
2 2
3 3
1 -1
1 1
2 2
3 3
3 4


# 1181번

알파벳 소문자로 이루어진 N개의 단어가 들어오면 아래와 같은 조건에 따라 정렬하는 프로그램을 작성하시오.

1. 길이가 짧은 것부터
2. 길이가 같으면 사전 순으로

In [14]:
from functools import cmp_to_key

def word_cmp(x, y):
  if (len(x) < len(y)):
    return -1
  elif (len(x) == len(y)):
    if (x < y):
      return -1
    elif (x == y):
      return 0
    else:
      return 1
  else:
    return 1


n_word = int(input())
words = []
for _ in range(n_word):
  words.append(input())

words = sorted(words, key = cmp_to_key(word_cmp))
word = ''
for w in words:
  if (word == w):
    continue
  else:
    word = w
    print(w)

3
i
no
no
i
no


# 10814번

온라인 저지에 가입한 사람들의 나이와 이름이 가입한 순서대로 주어진다. 이때, 회원들을 나이가 증가하는 순으로, 나이가 같으면 먼저 가입한 사람이 앞에 오는 순서로 정렬하는 프로그램을 작성하시오.

In [ ]:
n = int(input())
profile = []
for order in range(n):
  p = input().split()
  p.append(order)
  profile.append(tuple(p))

profile.sort(key=lambda x: (x[0], x[2]))

for person in profile:
  print(person[0], person[1])

# 18870번

수직선 위에 N개의 좌표 X1, X2, ..., XN이 있다. 이 좌표에 좌표 압축을 적용하려고 한다.

Xi를 좌표 압축한 결과 X'i의 값은 Xi > Xj를 만족하는 서로 다른 좌표의 개수와 같아야 한다.

X1, X2, ..., XN에 좌표 압축을 적용한 결과 X'1, X'2, ..., X'N를 출력해보자.

In [40]:
n = int(input())
arr = list(map(int, input().split()))
sorted_arr = sorted(list(set(arr)))

n_dict = {}
for i, n in enumerate(sorted_arr):
  if n in n_dict:
    continue
  else:
    n_dict[n] = i
  

answer = ''
for i in arr:
  answer += str(n_dict[i]) + ' '

print(answer[:-1])

6
10 9 10 9 10 9
1 0 1 0 1 0


In [41]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [42]:
cd /content/drive/MyDrive/git/baekjoon

/content/drive/MyDrive/git/baekjoon


In [43]:
!git config --global user.email 'jongmeoung8520@gmail.com'
!git config --global user.name 'dobbis'

!git add baekjoon.ipynb

!git commit -m 'complete level: sort'

!git push

[master 6b74b0e] complete level: sort
 1 file changed, 1 insertion(+), 1 deletion(-)
To https://github.com/dobbis/baekjoon.git
 ! [rejected]        master -> master (fetch first)
error: failed to push some refs to 'https://dobbis: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@github.com/dobbis/baekjoon.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.


In [45]:
!git pull

remote: Enumerating objects: 5, done.
remote: Counting objects: 100% (5/5), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (3/3), done.
From https://github.com/dobbis/baekjoon
   2f64512..321a57d  master     -> origin/master
hint: Waiting for your editor to close the file... error: unable to start editor 'editor'
Not committing merge; use 'git commit' to complete the merge.
